### DISCUSSION TITLE
Extension methods and DI

### DISCUSSION THREAD

**Discussion by josephpdunne** [microsoft/RulesEngine#102] [Q&A]
Hi,
I am using an Extension method for a rule which in turn needs to use data services.
Asp.Net core Api application with standard DI for services etc.
Some pseudo code
```c#
 RuleExtensions.Configure(_service); // needed by extension method);
 var reSettingsWithCustomTypes = new ReSettings { CustomTypes = new Type[] { typeof(RuleExtensions) } };
_rulesEngine = new RulesEngine.RulesEngine(workflowRules.ToArray(), _logger, reSettingsWithCustomTypes);
```
Having to call  RuleExtensions.Configure(_service) to add the scoped/transient service smells.
Is there any pattern to follow that I missed?
I am aware of service locator etc. and trying to avoid any anti-pattern.
One idea is a custom middleware method to set it but again, kinda smells.
Thanks,
Joe

**Comment by abbasc52** [comment]
Hi, I believe you are looking for using an instance of RuleExtensions class which can be initialised via DI.

If that is the case, you can pass an instance of RuleExtensions as ruleParameter.
```c#
//In you execution code
var ruleExtensionInstance = new RuleExtensions(_service);//or get it from DI

var extensionParam = new RuleParameter("ruleExtensions",ruleExtensionInstance);
var input1 = new RuleParameter("input1",input1);
var results = await rulesEngine.ExecuteAllRulesAsync("myWorkflow", extensionParam, input1);

```

Let me know if this answers your question.

**Reply by josephpdunne** [reply]
Hi,
Thanks very much for the reply.
I think I need to add some information.
The RuleExtensions component is my version of the Utils used for custom rules and is therefore a static class with static extension methods.
Because it is static I cannot instantiate it but it does use a scoped service.
It is this scoped service that I need to inject somehow.
I did try to add a instance constructor but the service must be stored in a static field so scoping and lifetime issues.
Can the rules engine use a instance rather than a static class for custom rules?
Thanks again.
Joe

**Reply by abbasc52** [reply]
Hi Joe,
Setting CustomTypes is more like importing these Types so that expression can use them.
```c#
var reSettingsWithCustomTypes = new ReSettings { CustomTypes = new Type[] { typeof(RuleExtensions) } };
```

Then if you want to use it as a static class, you can call the type by name
``` 
"Expression": "RuleExtensions.DoSomethingStatic()" 
```

If you want to use an instance of the Type, you need pass it as an input parameter when calling ExecuteAllRulesAsync
Assuming this is your code and you have a instance method called `DoSomethingInstance`.
```c#
//In you execution code
var ruleExtensionInstance = new RuleExtensions(_service);//or get it from DI

var extensionParam = new RuleParameter("ruleExtensions",ruleExtensionInstance);
var input1 = new RuleParameter("input1",input1);
var results = await rulesEngine.ExecuteAllRulesAsync("myWorkflow", extensionParam, input1);  //passing ruleExtension instance as an input
 ```
Then you can use the instance of RuleExtensions in rule as follows:
```
"Expression": "ruleExtensions.DoSomethingInstance()"
```

Hope it answers your question.

**Reply by josephpdunne** [reply]
Hi,
I get it now.
Pass any dependency in as a parameter when applying rules.
This works but I did decide to use the Middleware injection route reasoning that the extension is less coupled to usage.
I can provide more details if anyone is interested.
Thanks,
Joe

**Reply by tisdalestar6git** [reply]
hi Joe

Thanks for your post. I have been trying to use what you have modeled but I'm missing something, 
would you mind taking a look at what I have and tell me where I have gone wrong? or provide any assistance/guidance. 

Thanks


I'm getting this exception:
		ExceptionMessage = "Exception while parsing expression `ruleExtensions.RunMyRule()` - No applicable method 'RunMyRule' exists in type 'RuleExtensions'"


serice:

    public class testService
    {
           private readonly object _obj;

        public testService()          
        {
            _obj = new MessageSend();
        }

        public void RunMyRule()
        {
            string item =   "foobar";
        }

    }



Test rule json:
[
    {
        "WorkflowName": "DMEWorkflow",
        "Rules": [
            {
                "RuleName": "dmetest",
                "SuccessEvent": "10",
                "ErrorMessage": "dennisTesting.",
                "ErrorType": "Error",
                "RuleExpressionType": "LambdaExpression",
                "Expression": "ruleExtensions.RunMyRule()"
            }
        ]
    }
]


My test code:
			var ruleExtensionInstance = new RuleExtensions(_service);//or get it from DI

            var extensionParam = new RuleParameter("ruleExtensions", ruleExtensionInstance);
            var inputDI = new RuleParameter("input1", inputH);

            var dmefiles = Directory.GetFiles(Directory.GetCurrentDirectory(), "dmetest.json", SearchOption.AllDirectories);
            if (dmefiles == null || dmefiles.Length == 0)
                throw new Exception("Rules not found.");

            var DMEfileData = File.ReadAllText(dmefiles[0]);

            var DMEworkflowRules = JsonConvert.DeserializeObject<List<WorkflowRules>>(DMEfileData);

            var DMEbre = new RulesEngine.RulesEngine(DMEworkflowRules.ToArray(), null);


            List<RuleResultTree> resultListDME = DMEbre.ExecuteAllRulesAsync("DMEWorkflow", extensionParam, inputDI).Result;

**Reply by josephpdunne** [reply]
Hi, just a quick look but something to check. 
You are using an instance of the service class:
var ruleExtensionInstance = new RuleExtensions(_service);//or get it from DI

I think this has be a static class (only extension methods are callable from the expression) and should be:
var extensionParam = new RuleParameter("ruleExtensions", typeof(RuleExtensions) );
In the service class:

public static class testService
{
    private static object _obj;

    public testService()
    {
        // this is the problem,
        // no constructor for static class
        _obj = new MessageSend();
    }

    // hack
    public static object SetObject
    {
        set
        {
            _obj = value;
        }
    }
    public static void RunMyRule()
    {
        string item = "foobar";
    }

}

Before you call you have to set the _obj to an instance of SendMessage

var obj = new SendMessage();
RuleExtensions.SetObject(obj);
var DMEbre = new RulesEngine.RulesEngine(DMEworkflowRules.ToArray(), null);
... rest of code

What I am avoiding is that last bit, having to call RuleExtensions.SetObject(obj). 
I hide this by running equivalent code in the middleware.

I think this 
"Expression": "ruleExtensions.RunMyRule()"
should be:
"Expression": "RuleExtensions.RunMyRule()"

Hope that helps
Joe
